### PR TITLE
T1715

### PR DESCRIPTION
**Pull request by secretrobotron**

### CODE REVIEW

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_check** from @mjschranz received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/d5550941263c205/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/d5550941263c205/output.txt

Total script time: 0.41 mins

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @mjschranz received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/28e0ccbcc5a3909/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/28e0ccbcc5a3909/output.txt

Total script time: 1.06 mins

Server is running on http://pr640.butyr.org/templates/test.html

File: editors/googlemap-editor.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        // Update properties when TrackEvent is updated
+        trackEvent.listen( "trackeventupdated", function ( e ) {
+          _this.updatePropertiesFromManifest( e.target );
+          _targetSelectElement.value = trackEvent.popcornOptions.target;
+          setErrorState( false );
+        });
+        setup( trackEvent );
+      },
+      close: function () {
+      }
```
Comment (ScottDowne) [inline editors/googlemap-editor.js:252]: Cleaner like this:

close: function () {}

File: editors/googlemap-editor.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      if ( message ) {
+        _messageContainer.innerHTML = message;
+        _messageContainer.parentNode.style.height = _messageContainer.offsetHeight + "px";
+        _messageContainer.parentNode.style.visibility = "visible";
+        _messageContainer.parentNode.classList.add( "open" );
+      }
+      else {
+        _messageContainer.innerHTML = "";
+        _messageContainer.parentNode.style.height = "";
+        _messageContainer.parentNode.style.visibility = "";
```
Comment (ScottDowne) [inline editors/googlemap-editor.js:152]: Should set it to "visible" or "hidden" and not an empty string.

File: editors/googlemap-editor.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+     */
+    function setup( trackEvent ){
+      var targetList = _this.createTargetsList( _targets ),
+          optionsContainer = _rootElement.querySelector( ".editor-options" );
+
+      // Attach the onchange handler to trackEvent is updated when <select> is changed
+      _targetSelectElement = targetList.querySelector( "select" );
+
+      // Remove the default "Media Element" target so that the googlemaps plugin doesn't get angry with a blank target
+      _targetSelectElement.removeChild( _targetSelectElement.querySelector( '.default-target-option' ) );
```
Comment (ScottDowne) [inline editors/googlemap-editor.js]: Chnage these single quotes to double quotes.

File: src/editor/editor.js [hunk 4]
```diff
@@ -384,6 +384,43 @@ define( [ "core/eventmanager", "util/lang", "util/xhr",
...
+
+      if ( layoutsToLoad.length === 0 ) {
+        readyCallback();
+      }
+      else {
+        layoutsToLoad.forEach( function( editorHusk ) {
+          Editor.loadLayout( editorHusk.layout.substring( 5 ), function( layoutSrc ){
+            editorHusk.layout = layoutSrc;
+            ++loadedLayouts;
+            if ( loadedLayouts === layoutsToLoad.length ) {
```
Comment (ScottDowne) [inline src/editor/editor.js:415]: Could you remove the ++loadedLayouts and change the next line to:

if ( ++loadedLayouts === layoutsToLoad.length ) {

File: editors/googlemap-editor.js (hunk 2 above)
Comment (secretrobotron) [inline editors/googlemap-editor.js:152]: I do this on purpose, actually. It seemed like it was the only way to revert back to whatever it was before style was applied directly to the element, which means class-scope properties will work again.

File: src/editor/editor.js (hunk 4 above)
Comment (secretrobotron) [inline src/editor/editor.js:415]: I don't totally agree with this. The way it's written now, each line's intent is a lot more obvious for readers (which I like).

Comment (ScottDowne) [inline src/editor/editor.js:415]: Yeah, that's fair.

File: editors/googlemap-editor.js (hunk 2 above)
Comment (ScottDowne) [inline editors/googlemap-editor.js:152]: I did not know that. A trick I must remember.

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @ScottDowne received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/1708887fb2c7024/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_check** from @ScottDowne received. Current queue size: 1

Live output at: http://botio.butyr.org:8000/4524c57a288ddc7/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/1708887fb2c7024/output.txt

Total script time: 0.84 mins

Server is running on http://pr640.butyr.org/templates/test.html

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/4524c57a288ddc7/output.txt

Total script time: 0.43 mins

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @ScottDowne received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/4c8b16e8db109c7/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Failed

Full output at http://botio.butyr.org:8000/4c8b16e8db109c7/output.txt

Total script time: 0.62 mins

File: editors/googlemap-editor.js [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+        trackEvent.listen( "trackeventupdated", function ( e ) {
+          _this.updatePropertiesFromManifest( e.target );
+          _targetSelectElement.value = trackEvent.popcornOptions.target;
+          setErrorState( false );
+          getMapFromTrackEvent();
+        });
+        setup( trackEvent );
+      },
+      close: function () {
+        removeMapListeners();
```
Comment (k88hudson) [inline editors/googlemap-editor.js:251]: Do we really want to not save the position if the editor is closed? There's no way to indicate this distinction to the user, and I can't really see why we we wouldn't want to.

File: editors/googlemap-editor.js [hunk 6]
```diff
@@ -0,0 +1,257 @@
+/* This Source Code Form is subject to the terms of the MIT license
+ * If a copy of the MIT license was not distributed with this file, you can
+ * obtain one at http://www.mozillapopcorn.org/butter-license.txt */
+
+( function( Butter ) {
+
+  Butter.Editor.register( "googlemap", "load!{{baseDir}}editors/googlemap-editor.html", function( rootElement, butter ) {
```
Comment (k88hudson) [inline editors/googlemap-editor.js:7]: Is it really necessary to load html in here?

Comment (secretrobotron) [inline editors/googlemap-editor.js:7]: Well, strictly speaking, we could just write it all out in javascript, but I'd rather not. This is really an example of how to use this system for building more complex editors that do require an html layout to be loaded.

File: editors/googlemap-editor.js (hunk 5 above)
Comment (secretrobotron) [inline editors/googlemap-editor.js:251]: How does a user close an editor now? They can't manually, really. If they change any attribute in the editor, a save happens, and if they change the map, a save also happens. How else can they change something without a save happening immediately afterward?

File: templates/complete-config.json [hunk 7]
```diff
@@ -3,6 +3,7 @@
   "title": "Complete",
   "baseDir": "../",
   "editor": {
+    "googlemap": "{{baseDir}}editors/googlemap-editor.js"
```
Comment (jbuck) [inline templates/complete-config.json:6]: Can we put this in the default config so that it's accessible to all templates?

File: src/modules.js [hunk 8]
```diff
@@ -14,7 +14,7 @@ define(
 
   var moduleList = Array.prototype.slice.apply( arguments );
 
-  return function( butter, config, onReady ){
+  return function( Butter, butter, config, onReady ){
```
Comment (jbuck) [inline src/modules.js:17]: Why the case change here?

File: src/main.js [hunk 9]
```diff
@@ -760,7 +760,7 @@
         _this.project.template = _config.value( "name" );
 
         //prepare modules first
-        var moduleCollection = Modules( _this, _config ),
+        var moduleCollection = Modules( Butter, _this, _config ),
```
Comment (jbuck) [inline src/main.js:763]: Why the case change and added parameter here? It doesn't look like it's being used here.

File: editors/googlemap-editor.js (hunk 6 above)
Comment (jbuck) [inline editors/googlemap-editor.js:7]: Any reason to not use the existing requirejs text plugin to load HTML?

File: editors/googlemap-editor.js [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+    var _rootElement = rootElement,
+        _targets = [ butter.currentMedia ].concat( butter.targets ),
+        _messageContainer = _rootElement.querySelector( "div.error-message" ),
+        _targetSelectElement,
+        _trackEvent,
+        _popcornEventMapReference,
+        _mapListeners;
+
+    /**
+     * Member: onCenterChanged
```
Comment (jbuck) [inline editors/googlemap-editor.js:20]: Wrong method name here

**Comment by jbuck** [comment]
Staged on master: https://github.com/mozilla/butter/commit/59fdcc1b0a871611ee150584c740711ea1480eb0

**Comment by jbuck** [comment]
Staged on master: https://github.com/mozilla/butter/commit/59fdcc1b0a871611ee150584c740711ea1480eb0

**Pull request closed without merging**